src/contract/event.rs

```diff
@@ -293,27 +296,58 @@ impl<T: Transport, E: Detokenize> Stream for EventStream<T, E> {
     }
 }
 
-/// Trait for parsing a transaction log into an some event data when the
-/// expected event type is not known.
-pub trait TryFromLog: Sized {
-    fn from_log(log: Log) -> Result<Self, ExecutionError>;
-}
-
-/// Raw log transaction data.
+/// Raw log topics and data for a contract event.
 #[derive(Clone, Debug, Eq, PartialEq)]
-pub struct RawEventData {
-    /// The raw topics.
+pub struct RawLog {
+    /// The raw 32-byte topics.
     pub topics: Vec<H256>,
-    /// Encoded log data.
+    /// The raw non-indexed data attached to an event.
     pub data: Vec<u8>,
 }
 
-impl TryFromLog for RawEventData {
-    fn from_log(log: Log) -> Result<Self, ExecutionError> {
-        Ok(RawEventData {
+impl RawLog {
+    /// Decode raw log data into a tokenizable for a matching event ABI entry.
+    pub fn decode<D>(self, event: &AbiEvent) -> Result<D, ExecutionError>
+    where
+        D: Detokenize,
+    {
+        let event_log = event.parse_log(AbiRawLog {
+            topics: self.topics,
+            data: self.data,
+        })?;
+
+        let tokens = event_log
+            .params
+            .into_iter()
+            .map(|param| param.value)
+            .collect::<Vec<_>>()
+            .compat()
+            .ok_or(ExecutionError::UnsupportedToken)?;
+        let data = D::from_tokens(tokens)?;
+
+        Ok(data)
+    }
+}
+
+impl From<Log> for RawLog {
+    fn from(log: Log) -> Self {
+        RawLog {
             topics: log.topics,
             data: log.data.0,
-        })
+        }
+    }
+}
+
+/// Trait for parsing a transaction log into an some event data when the
+/// expected event type is not known.
+pub trait TryFromLog: Sized {
```

Comment:
why are you defining your own trait here? can't we `use std::convert::TryFrom<RawLog>`? Also why do we return a Result if it's always Ok?

Reply:
I will rename it. It should be `ParseLog` or something. This is implemented by the generated events enum. While parsing a raw log into a raw log can never fail, parsing a raw log into a generated events enum (if it doesn't match any of the variants for example), so it needs to return a `Result`.